### ISSUE TITLE
Null value for nested document causes TypeError when creating new document

### ISSUE THREAD

**Issue by silentjohnny** [Automattic/mongoose#9880]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

When a document is retrieved from DB that has a nested document with a null value, assigning that value to a property of a new, empty document causes a TypeError

```
TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at model.$set (/Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/document.js:944:19)
    at model.$set (/Users/joris/Developer/Projects/mongoose-test//node_modules/mongoose/lib/document.js:990:14)
    at model.Document (/Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/document.js:150:12)
    at model.Model (/Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/model.js:105:12)
    at new model (/Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/model.js:4706:15)
    at /Users/joris/Developer/Projects/mongoose-test/src/scripts/test-null-values.js:148:25
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/model.js:4863:16
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/model.js:4863:16
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/helpers/promiseOrCallback.js:24:16
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/model.js:4886:21
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/mongoose/lib/query.js:4397:11
    at /Users/joris/Developer/Projects/mongoose-test/node_modules/kareem/index.js:136:16
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
```

This is caused by a change made in 5.10.16 for #9563 

in document.js, line 927:

```
 if (path == null) {
    const _ = path;
    path = val;
    val = _;
  } else if (typeof path !== 'string') {
    // new Document({ key: val })
    if (path instanceof Document) {
      if (path.$__isNested) {
        path = path.toObject();
      } else {
        path = path._doc;
      }
    }

    prefix = val ? val + '.' : '';

    keys = Object.keys(path);
```

here, the check for path == null was moved up, but the value of path actually is a nested Document with a null value, so 

```
path.toObject()
```

will result in a null value and then `Object.keys(path)` will throw a TypeError

**If the current behavior is a bug, please provide the steps to reproduce.**

```
const mongoose = require('mongoose');
const { Schema } = mongoose;

const testSchema = new Schema({
  prop:  String,
  nestedProp: {
    prop: String
  }
});

mongoose.connect('mongodb://localhost/test', {useNewUrlParser: true, useUnifiedTopology: true});

const Test = mongoose.model('Test', testSchema);

new Test({
  prop: 'Test',
  nestedProp: null
}).save((err, doc) => {
  console.log(doc.id)
  console.log(doc.nestedProp)

  // let's clone this document:
  const clone = new Test({
    prop: 'Test 2',
    nestedProp: doc.nestedProp
  })

  // so far, so good:
  console.log(clone.id) // 'cloned document id'
  console.log(clone.nestedProp) // '{}'

  // let's update the document:
  Test.updateOne({
    _id: doc._id
  }, {
    nestedProp: null
  }, (err) => {

    // ... and retrieve it
    Test.findOne({
      _id: doc._id
    }, (err, updatedDoc) => {

      // now, this is interesting:
      console.log(updatedDoc.id) // 'document id'
      console.log(updatedDoc.nestedProp) // 'MongooseDocument { null }'

      // now this will throw a TypeError:
      const failing = new Test({
        prop: 'Test 3',
        nestedProp: updatedDoc.nestedProp
      })

    })
  })
})
```

When saving the document, the actual MongoDB does not contain the `nestedProp`
So this only happens if the document is updated with a null value, in which case the actual MongoDB document contains a `null` value. 

In other words, looking at the code example above:

`originalDoc.nestedProp.toObject()` evaluates to `{}`
`updatedDoc.nestedProp.toObject()` evaluates to `null`

**What is the expected behavior?**

The value of path should be recognized as null and set the property in the new document to null

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**

Using Node 14.15.3, Mongoose 5.11.14 and MongoDB 4.2

**Comment by silentjohnny** [comment]
One more thing: downgrading to Mongoose 5.10.15 solves the issue

**Comment by IslandRhythms** [comment]
I'm using mongodb@3.6.3 and node@15.7 with the same mongoose version and getting the same problem
